pkg/cache/internal/informers_map.go

```diff
@@ -162,9 +164,19 @@ func (ip *specificInformersMap) Get(gvk schema.GroupVersionKind, obj runtime.Obj
 	}
 
 	if started && !i.Informer.HasSynced() {
+		syncReturn := make(chan bool, 1 /* don't leak goroutines on timeout */)
+		go func() {
+			syncReturn <- cache.WaitForCacheSync(ctx.Done(), i.Informer.HasSynced)
+		}()
+
 		// Wait for it to sync before returning the Informer so that folks don't read from a stale cache.
-		if !cache.WaitForCacheSync(ip.stop, i.Informer.HasSynced) {
-			return started, nil, fmt.Errorf("failed waiting for %T Informer to sync", obj)
+		select {
+		case <-ctx.Done():
+			return started, nil, apierrors.NewTimeoutError(fmt.Sprintf("timeout waiting for %T Informer to sync", obj), 0)
+		case syncSuccess := <-syncReturn:
+			if !syncSuccess {
+				return started, nil, fmt.Errorf("failed waiting for %T Informer to sync", obj)
+			}
```

Comment:
Is there a race condition here between `cache.WaitForCacheSync()` and the `<-ctx.Done()` select case?

When `ctx.Done()` is closed, `cache.WaitForCacheSync()` returns false, and we send that boolean value to the `syncReturn` channel. That effectively ties `ctx.Done()` to `syncReturn`, which means our select block cases are waiting for the same condition.

Could this be simplified to the following?
```go
canceled := cache.WaitForCacheSync(ctx.Done(), i.Informer.HasSynced)
if canceled {
	return started, nil, apierrors.NewTimeoutError(fmt.Sprintf("timeout waiting for %T Informer to sync", obj), 0)
}
```

Reply:
I implemented this. What you said makes sense.